cmd/snapshot/main.go

```diff
@@ -0,0 +1,183 @@
+	if len(bounds) != 4 {
+		fmt.Fprintf(os.Stderr, "Error: invalid bounds provided — %v\n", boundString)
+		usage()
+		os.Exit(2)
+	}
+	for i, bound := range bounds {
+		FBounds[i], err = strconv.ParseFloat(strings.TrimSpace(bound), 64)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: invalid bounds provided — %v\n", boundString)
+			fmt.Fprintf(os.Stderr, "Error: unabled to parse %v(%v) as a float.\n", bound, i)
+			usage()
+			os.Exit(2)
+		}
+	}
+}
+
+func parseTile(tileString string) {
+	var err error
+	var v uint64
+	parts := strings.Split(tileString, "/")
+	if len(parts) != 3 {
+		fmt.Fprintf(os.Stderr, "Error: invalid z/x/y coordinates  — %v\n", tileString)
+		usage()
+		os.Exit(2)
+	}
+	var label = [...]string{"Z", "X", "Y"}
+
+	for i, part := range parts {
+		switch i {
+		case 0:
+			v, err = strconv.ParseUint(strings.TrimSpace(part), 10, 64)
+			FTile.Z = uint(v)
+
+		case 1:
+			v, err = strconv.ParseUint(strings.TrimSpace(part), 10, 64)
+			FTile.X = uint(v)
+		case 2:
+			v, err = strconv.ParseUint(strings.TrimSpace(part), 10, 64)
+			FTile.Y = uint(v)
+		}
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: invalid tile coordinates provided — %v\n", tileString)
+			fmt.Fprintf(os.Stderr, "Error: unabled to parse %v %v as a uint.\n", label[i], part)
+			usage()
+			os.Exit(2)
+		}
+	}
+}
+
+func ParseFlags() cmdType {
+
+	flag.UintVar(&FWidth, "width", 100, "Width of the image to generate.")
+	flag.UintVar(&FWidth, "w", 100, "Width of the image to generate.")
+
+	flag.UintVar(&FHeight, "height", 100, "Height of the image to generate.")
+	flag.UintVar(&FHeight, "h", 100, "Height of the image to generate.")
+
+	flag.StringVar(&FStyle, "style", "https://osm.tegola.io/capabilities/osm.json", "Style file")
+	flag.Float64Var(&FPixelRatio, "pixel", 1.0, "The pixel ratio")
+
+	flag.Parse()
+
+	if flag.NArg() < 3 {
+		usage()
+		os.Exit(2)
+	}
+
+	var cmd cmdType
+	var fileIdx = 2
+
+	switch strings.TrimSpace(strings.ToLower(flag.Arg(0))) {
+	case "bounds":
+		// The next variable should be the bounds seperated by spaces.
+		parseBounds(flag.Arg(1))
+		cmd = CmdBounds
+	case "tile":
+		// The next variable should be the coordinates sepearted by forward slash.
+		parseTile(flag.Arg(1))
+
+		cmd = CmdTile
+	default: // assume bounds as the default subcommand
+		parseBounds(flag.Arg(0))
+		fileIdx = 1
+		cmd = CmdBounds
+	}
+
+	// Next should be the output filename.
+	FOutputFilename = strings.TrimSpace(flag.Arg(fileIdx))
+	return cmd
+
+}
+
+func main() {
+	var file *os.File
+	var err error
+	var img image.Image
+	cmd := ParseFlags()
+	file, err = os.Create(FOutputFilename)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error creating output file: %v\n", err)
+		os.Exit(3)
+	}
+	ss := mbgl.NewSnapshotter(FStyle, float32(FPixelRatio))
+	size := image.Point{
+		X: int(FWidth),
+		Y: int(FHeight),
+	}
+	switch cmd {
+	case CmdBounds:
+		img = ss.Snapshot(&FBounds, size)
+
+	case CmdTile:
+		img = mbgl.SnapshotTile(ss, FTile, size)
```

Comment:
@ear7h I think instead of size, which is difficult to understand what is going on here. Since you are setting the X, Y for width and height. I think instead you should just take a `width` and `height` parameter of type uint.

Reply:
Also, these should return errors.

Reply:
Errors need to be implemented on the C side. #1 is tracking this and I currently put in a suggestion for this using macros